lib/ex_plasma/transaction/signed.ex

```diff
@@ -0,0 +1,124 @@
+defmodule ExPlasma.Transaction.Signed do
+  @moduledoc """
+  Representation of a signed transaction.
+
+  NOTE: before you use this, make sure you shouldn't use `Transaction` or `Transaction.Recovered`
+  """
+
+  alias ExPlasma.Crypto
+  alias ExPlasma.Transaction
+  alias ExPlasma.Transaction.Protocol
+  alias ExPlasma.Transaction.Witness
+  alias ExPlasma.TypedData
+  alias ExPlasma.Utils.RlpDecoder
+
+  @type tx_bytes() :: binary()
+  @type decoding_error() :: :malformed_rlp | mapping_error()
+
+  @type mapping_error() ::
+          :malformed_transaction
+          | :malformed_witnesses
+          | atom()
+
+  @type validation_error() :: {:witnesses, :malformed_witnesses} | {atom(), atom()}
+
+  @type t() :: %__MODULE__{
+          raw_tx: Protocol.t(),
+          sigs: [Crypto.sig_t()]
+        }
+
+  defstruct [:raw_tx, :sigs]
+
+  @doc """
+  Produce a binary form of a signed transaction - coerces into RLP-encodeable structure and RLP encodes
+  """
+  @spec encode(t()) :: tx_bytes()
+  def encode(%__MODULE__{} = signed) do
+    signed |> to_rlp() |> ExRLP.encode()
+  end
+
+  @doc """
+  RLP encodes the signed transaction.
+  """
+  @spec to_rlp(t()) :: list()
+  def to_rlp(%__MODULE__{} = signed) do
+    [signed.sigs | Protocol.to_rlp(signed.raw_tx)]
+  end
+
+  @doc """
+  Produces a struct from the binary encoded form of a signed transactions - RLP decodes to structure of RLP-items
+  and then produces an Elixir struct.
+
+  Only validates that the RLP is structurally correct.
+  Does not perform any other kind of validation, use validate/1 for that.
+  """
+  @spec decode(tx_bytes()) :: {:ok, t()} | {:error, decoding_error()}
+  def decode(signed_tx_bytes) do
+    with {:ok, tx_rlp_decoded_chunks} <- RlpDecoder.decode(signed_tx_bytes) do
+      to_map(tx_rlp_decoded_chunks)
+    end
```

Comment:
with's should be reserved for multiple statement chaining. use a `case` instead.